packages/server/modules/gatekeeper/services/subscriptions/upgradeWorkspaceSubscription.ts

```diff
@@ -0,0 +1,381 @@
+import {
+  GetWorkspacePlan,
+  GetWorkspacePlanPriceId,
+  GetWorkspacePlanProductId,
+  GetWorkspaceSubscription,
+  ReconcileSubscriptionData,
+  SubscriptionDataInput,
+  UpsertPaidWorkspacePlan,
+  UpsertWorkspaceSubscription
+} from '@/modules/gatekeeper/domain/billing'
+import { CountSeatsByTypeInWorkspace } from '@/modules/gatekeeper/domain/operations'
+import {
+  InvalidWorkspacePlanUpgradeError,
+  WorkspaceNotPaidPlanError,
+  WorkspacePlanMismatchError,
+  WorkspacePlanNotFoundError,
+  WorkspacePlanUpgradeError,
+  WorkspaceSubscriptionNotFoundError
+} from '@/modules/gatekeeper/errors/billing'
+import { isNewPlanType, isOldPaidPlanType } from '@/modules/gatekeeper/helpers/plans'
+import { calculateNewBillingCycleEnd } from '@/modules/gatekeeper/services/subscriptions/calculateNewBillingCycleEnd'
+import { mutateSubscriptionDataWithNewValidSeatNumbers } from '@/modules/gatekeeper/services/subscriptions/mutateSubscriptionDataWithNewValidSeatNumbers'
+import { isUpgradeWorkspacePlanValid } from '@/modules/gatekeeper/services/upgrades'
+import { NotImplementedError } from '@/modules/shared/errors'
+import { CountWorkspaceRoleWithOptionalProjectRole } from '@/modules/workspaces/domain/operations'
+import {
+  PaidWorkspacePlans,
+  PaidWorkspacePlansNew,
+  throwUncoveredError,
+  WorkspacePlanBillingIntervals,
+  xor
+} from '@speckle/shared'
+import { cloneDeep } from 'lodash'
+
+export const upgradeWorkspaceSubscriptionFactoryOld =
```

Comment:
This one is a copy/paste from main

Reply:
another downside of copy pasting the new implementation - very hard to review. i have to keep both the old and new versions side by side and go line by line to see what actually differs between them and how

Reply:
The point is that you should not compare the 2, the old one is exactly as it was before.
The new implementation is similar but that's not the rule, you should review the new one on its own.

Reply:
> The point is that you should not compare the 2
I disagree. Especially as I don't have the full context & experience with this function, it is actually very important for me to compare it with a baseline - the previous implementation